### DISCUSSION TITLE
Revisiting Taxonomy design for shared structures (like Connection, Action, etc )

### DISCUSSION THREAD

**Discussion by rohithdv** [fybrik/fybrik#1695] [Developers]
The fybrik manager is currently following the [below taxonomy structure](https://github.com/fybrik/fybrik/blob/8acde22beb834cac02087652f4ea24c21beceba4/charts/fybrik/files/taxonomy/taxonomy.json#L37) (in json schema format) for sending asset `Connection` Information to the connectors.  
```
"Connection": {
    "type": "object",
    "required": [
      "name"
    ],
    "properties": {
      "name": {
        "$ref": "#/definitions/ConnectionType"
      }
    },
    "additionalProperties": true
},
"ConnectionType": {
   "type": "string"
}
```
In summary, by current design, a `Connection` structure should have a `name` field followed by ANY `map[string]interface{}` field (since `additionalProperties` is set to true). 

So, to be compliant with the above format, fybrik manager is currently constructing `Connection` information (during the create / copy request) using the below[ golang code snippet](https://github.com/fybrik/fybrik/blob/master/manager/controllers/app/plotter_generator.go#L76).
```
connection := taxonomy.Connection{
    Name: cType,
    AdditionalProperties: serde.Properties{
      Items: map[string]interface{}{
        string(cType): map[string]interface{}{
          "endpoint":   bucket.Endpoint,
          "bucket":     bucket.Name,
          "object_key": genObjectKeyName,
        },
      },
    },
  }
```
Where `cType` can be “s3”, “db2”, etc. 
An example Connection request sent by manager today (conforming to the above format) is given below. 
```
"connection": {
   "name": "s3",
   "s3": {
      "endpoint": "http://localstack.fybrik-notebook-sample.svc.cluster.local:4566/",
      "bucket": "demo",
      "object_key": "PS_20174392719_1491204439457_log.csv"
   }
}
```
If the corresponding connectors are also following the above connection structure to deserialise the connection information sent by manager then it is ok. 

However, if the connectors decide to have a different taxonomy structure for receiving the asset `Connection` information (which, by  design, is allowed for the connectors to have their own taxonomy), then we have an issue. 

For eg: If the [connectors change](https://github.ibm.com/IBM-Data-Fabric/WKC-connector/blob/1e03ce9c30fa2012d45dfe15bc1b3ff79fca511d/src/main/resources/taxonomy/taxonomy.json#L112) the `Connection` structure (in json schema format) to the following: 
```
"Connection": {
      "type": "object",
      "description": "Connection information",
      "properties": {
        "name": {
          "$ref": "#/definitions/ConnectionName"
        },
        "additionalProperties": {
          "$ref": "#/definitions/AdditionalProperties"
        }
      },
      "additionalProperties": false,
      "required": [
        "name"
      ]
    },
"AdditionalProperties": {
      "type": "object",
      "description": "Connection information",
      "properties": {
        "datasourceType": {
          "type": "string"
        },
        "assetProperties": {
          "$ref": "#/definitions/AssetProperties"
        },
        "columns": {
          "$ref": "#/definitions/Columns"
        },
        "connectionString": {
          "type": "string"
        },
        "connectionProperties": {
          "$ref": "#/definitions/ConnectionProperties"
        }
      },
      "required": [
        "assetProperties",
        "columns",
        "connectionProperties"
      ]
    },
    "AssetProperties": {
      "type": "array",
      "description": "asset connection information",
      "items": {
        "$ref": "#/definitions/AssetDetail"
      }
    },
    "AssetDetail": {
      "type": "object",
      "description": "asset connection information",
      "properties": {
        "name": {
          "type": "string"
        },
        "value": {
          "type": "string"
        }
      },
      "required": [
        "name",
        "value"
      ]
    },
```
An example of a Connection object conformant to the above new format is  given below: 
```
"connection": {
    "name": "catalog",
    "additionalProperties": {
      "assetProperties": [],
      "columns": [],
      "connectionProperties": {
       "name": "s3",
       "s3": {
          "bucket": "...",
          "endpoint": "...",
          "object_key": "..."
        }
      },
   }
}
```
Now, there is a mismatch between the `Connection` formats of the manager and the connectors. This leads to failure in deserialisation of the `Connection` structure on the connector side since manager is not able to send the new fields like AdditionalProperties, AssetProperties. etc.

**Comment by rohithdv** [comment]
One approach is the following: 
1. Store the manager `taxonomy.json` (and any other relevant manager taxonomy files) in the connectors in a separate folder.
2. Generate code in the programming language used in the connector (in a folder called `managerTaxonomy`) during connector build stage corresponding to the above taxonomy.
3. Create a new API as part of the Manager-Connector REST API interface. 
4. Let this API be called  `getConnectorTaxonomyConnection(in *managerTaxonomy.Connection) (managerTaxonomy.Connection, error)`
5. The purpose of `getConnectorTaxonomyConnection` API is to convert the `Connection` structure from the `manager taxonomy `format to the `connectors taxonomy` format and return the `connectors taxonomy compliant connection` as a `managerTaxonomy.Connection` object.
6.  The `getConnectorTaxonomyConnection` API will understand the manager taxonomy format and connectors taxonomy format and can convert between the two formats. 
7. Manager will need to call the `getConnectorTaxonomyConnection` API before it calls [CreateAsset](https://github.com/fybrik/fybrik/blob/8acde22beb834cac02087652f4ea24c21beceba4/pkg/connectors/datacatalog/clients/datacatalog.go#L15) or any API in which it needs to construct a `Connection` object. 
8. `getConnectorTaxonomyConnection` API return the `Connection` structure compliant with latest connectors taxonomy.
9. Now manager uses the returned `managertaxonomy.Connection` structure and constructs the CreateAssetRequest and calls the `CreateAsset` API (or any other API using the `Connection` structure) on the connectors. 

The rationale behind the above approach is that any changes to the connectors taxonomy will be handled by the corresponding connector by using information about the taxonomy understood by the manager.

**Reply by shlomitk1** [reply]
Just to be aware of: there is another use of the Connection construction - in modules. Whenever a temporary copy is created, or "in-memory" connection is used, the manager constructs a new Connection and sends to modules. The API-based connection (e.g. virtual endpoint defined in FybrikApplication status) is done based on the module API definition in yamls, so no "hard-coded" structure is created.
So, it seems that the only "hard-coding" is done for storage allocation. I wonder if we can change FybrikStorageAccount definition for defining the Connection json string with place-holders for %bucket and %object values that will be filled in by manager. This way, if tomorrow we add DB2 storage allocation, the corresponding FybrikStorageAccount will hold the relevant information for DB2 connection, and so on.

**Reply by rohithdv** [reply]
But how will this placeholder approach resolve the different formats issue ? The idea is that the manager need not be aware of the structure information ( or taxonomy) used in connectors or modules. Tomorrow one connector/module may have one connection structure and another connector/ module may have different connection structure. It may be difficult for the manager to be aware of these specific structural changes. I think there needs to   be some conversion mechanism so that the manager can construct the right structure.

**Reply by shlomitk1** [reply]
The placeholder approach makes the manager unaware of the connection taxonomy. The idea is still the same: modules and connectors share the same taxonomy, the manager is unaware of the taxonomy. 
The storage account will have the ready connection object, properly configured. The IT admin will still have to do the configuration inside yamls but it won't be hard-coded within manager.

**Reply by shlomitk1** [reply]
Example of the StorageAccount yaml (according to the first taxonomy sample, the connection field will be different for the second sample)
...
```
"region": "regionA",
"endpoint": "http://localstack.fybrik-notebook-sample.svc.cluster.local:4566/",
"secretRef": "my-secret",
"connection": '{
   "name": "s3",
   "s3": '{
      "endpoint": "http://localstack.fybrik-notebook-sample.svc.cluster.local:4566/",
      "bucket": .{{ .Bucket }}
      "object_key": .{{ .Object }}
   }
}'
```

**Reply by rohithdv** [reply]
Ok got it. Thanks. So in the placeholder approach, we will need to put the relevant connector specific storage account yamls near to the manager. So the updated  connection structure info can be sent to both the relevant connectors and modules. 

In the API based approach, we need to code in the connector specific connection information in the connectors. The manager will call the connectors to get the updated connection structure. However in this approach how the manager will convey the updated connection structure to the modules is a question still.

**Reply by simanadler** [reply]
> Just to be aware of: there is another use of the Connection construction - in modules. Whenever a temporary copy is created, or "in-memory" connection is used, the manager constructs a new Connection and sends to modules. The API-based connection (e.g. virtual endpoint defined in FybrikApplication status) is done based on the module API definition in yamls, so no "hard-coded" structure is created. So, it seems that the only "hard-coding" is done for storage allocation. I wonder if we can change FybrikStorageAccount definition for defining the Connection json string with place-holders for %bucket and %object values that will be filled in by manager. This way, if tomorrow we add DB2 storage allocation, the corresponding FybrikStorageAccount will hold the relevant information for DB2 connection, and so on.

Yes, I was going to suggest the same thing.

**Comment by roytman** [comment]
Hi, thanks for the answers, maybe because I'm relatively new in the taxonomy world, I have several questions:

1. @rohithdv in your approach, the call `getConnectorTaxonomyConnection(in *managerTaxonomy.Connection) (managerTaxonomy.Connection, error)` gets and returns the same object type `managerTaxonomy.Connection`. Should it return `connectorTaxonomy.Connection` ?
2. why manager should call it before [CreateAsset](https://github.com/fybrik/fybrik/blob/8acde22beb834cac02087652f4ea24c21beceba4/pkg/connectors/datacatalog/clients/datacatalog.go#L15) ? Manager can send use its `managerTaxonomy.Connection` and the connector will translate it to the appropriate internal object.
3. A general taxonomy question, one of issues addressed by taxonomy, is 

> Dynamic Updates: New terms and new values can be added, removed and updated dynamically. For example, one can add new enforcement actions, new connection types, new purposes, etc without needing to redeploy fybrik.

I doubt that we can add new connection types. Ones, fybrik uses knowledge about internal connection fields like "bucket name", "bucket endpoint" it is not transparent. Today, we allocate new storage of the S3 type only. Let's try to do it for MySQL or DB2, it will require different parameters. An if tomorrow we will need to support a completely different connection type, how will manager handle it?

4. Now if we return to the "placeholder" suggestion, maybe we need not only structure definition, but a library with helper functions, e.g. allocate new storage, release an old storage asset ..... For each supported connection/dataAssetType, we create goLoang [plugin](https://pkg.go.dev/plugin) library. These plugins will be loaded during start of manager, they can be part of the same image, or mounted similar to the optimizer. In order to add a new connection type, we'll have to restart manager (goLang doesn't allow dynamically reload plugins), but in production HA deployments it can be done by a rolling update. 
It is not clear how we can add this approach to the modules/connectors which are written in different languages. However, at least for modules, it should not be a big problem, because a module supports a specific predefined connection type(s), and modules are dynamically deployed.

**Reply by rohithdv** [reply]
Response for 1. The main idea in the API based approach was manager does not need to know connector taxonomy. But the connectors need to know the manager taxonomy. So in the API definition l have used only the manager taxonomy for both the input parameter as well as return type. The manager does not have knowledge of the connector taxonomy. Hence it can't understand if the return type is connector taxonomy.

In the manager taxonomy connection struct we have 2 fields. `Name` which is a string and `additionalProperties` which is a `map[string]interface{}`. In the input parameter of the api the manager will send its understanding of the format of connection ( for example as given below).
```
"connection": {
	"name": "s3",
	"s3": {
		"endpoint": "http://localstack.fybrik-notebook-sample.svc.cluster.local:4566/",
		"bucket": "demo",
		"object_key": "PS_20174392719_1491204439457_log.csv"
	}
}
```

After the api call returns it will get back the same information however this time the `additionalProperties` field which is also a `map[string]interface{}` but will now have the `connector specific taxonomy` information. For example: the response for the above example would look something like this: 

```
"connection": {
	"name": “wkc”,
	"additionalProperties": {
		"assetProperties": [],
		"columns": [],
		"connectionProperties": {
			"name": "s3",
			"s3": {
				"bucket": "demo",
				"endpoint": "http://localstack.fybrik-notebook-sample.svc.cluster.local:4566",
				"object_key": "PS_20174392719_1491204439457_log.csv"
			}
		},
	}
},
```
The above can also be parsed as a `managerTaxonomy.Connection` as the connector's format of connection should also be compliant with manager's view of connection. Hence the above can be viewed as an example of `managerTaxonomy.Connection `and hence can be parsed by the manager.

2. What you have mentioned in Point 2 is exactly the way we do it today. The manager sends connection information as in the below format.
```
"connection": {
	"name": "s3",
	"s3": {
		"endpoint": "http://localstack.fybrik-notebook-sample.svc.cluster.local:4566/",
		"bucket": "demo",
		"object_key": "PS_20174392719_1491204439457_log.csv"
	}
}
```
But the connector cannot understand this as it expects connection something like below. 
```
"connection": {
	"name": “wkc”,
	"additionalProperties": {
		"assetProperties": [],
		"columns": [],
		"connectionProperties": {
			"name": "s3",
			"s3": {
				"bucket": "demo",
				"endpoint": "http://localstack.fybrik-notebook-sample.svc.cluster.local:4566",
				"object_key": "PS_20174392719_1491204439457_log.csv"
			}
		},
	}
},
```
The above structure is known only to the connector and not the manager.

 When we call `CreateAsset` today, the `Connection` struct by the manager sent in the `CreateAsset` is parsed on the connector side as a `connectorTaxonomy.Connection` and it is failing now as it does not have the required new fields like assetProperties, etc as defined according to the latest connector taxonomy. 

3. Taxonomy was designed to address these type of issues only. Basically, the purpose of taxonomy was to provide a common place to define various terms and variables which can be understood by all components (connectors, modules, etc) of the fybrik eco-system. 

In my understanding, the only difference between the situation we thought of earlier and the situation we are facing now is regarding the role of manager. 

In our current design of the taxonomy, the manager did not worry about the fields of the connection type (s3, db2, etc) . It just viewed all the specific fields of a connection type to be encoded in  a `map[string]interface{}`. 
Now, we have a scenario where manager needs to be aware of the particular fields (of s3, db2, etc) as it needs to construct the `Connection` object itself and this construction needs to be understood by other `connectors/modules`. 

One way we can support multiple connection types is by editing the `taxonomy.json` of the manager to provide connection specific variables. (Right now we dont have these type of information in the manager` taxonomy.json`). For example, for `s3` it could be something like this. 
```
"s3": {
      "type": "object",
      "description": "Connection information for S3 compatible object store",
      "properties": {
        "bucket": {
          "type": "string"
        },
        "endpoint": {
          "type": "string"
        },
        "object_key": {
          "type": "string"
        },
        "region": {
          "type": "string"
        },
        "api_key": {
          "type": "string"
        },
        "resource_instance_id": {
          "type": "string"
        },
        "access_key": {
          "type": "string"
        },
        "secret_key": {
          "type": "string"
        }
      },
      "required": [
        "bucket",
        "endpoint",
        "object_key"
      ]
    }
```
and all the `connectors/modules` need to follow this part of the taxonomy for seamless communication between each other. 

Currently these type of connection definitions are defined in the connectors taxonomy. If manager needs to know this type of information now, we should move these type of connection type information to the manager taxonomy and ask all other components to be conformant with this taxonomy. That way we can ensure everyone understands these connection specific terms. 

4. I will leave the response for this comment to @shlomitk1 . 
@roytman

**Comment by shlomitk1** [comment]
https://github.com/fybrik/fybrik/issues/1737 suggests among the rest to take out the connection creation from the manager. We are talking about an independent component that is capable of creating a new connection (while allocating storage if required). The connection object will be of a pre-configured form including placeholders like {{ .Values.endpoint }} and similar. If customers come out with a connection that is not supported by our open-source version of the storage allocator, they will have to provide their own version.

**Comment by simanadler** [comment]
The question is who/which component in the architecture "owns" the definition of the types of connections supported.  I think that the most natural answer is that the data catalog defines this.  Modules and connectors are then required to support a subset of what is supported by the data catalog.  The manager should not have any hard coded dependencies on specific types of connections as it does today for write and copy flows.  This is all very related to issue #1737.  Once we address that hopefully the taxonomy issues described here will more or less be addressed.  @shlomitk1 do you agree?

**Reply by shlomitk1** [reply]
Agree with @simanadler that the data catalog defines the supported connections, and the manager should not have hard-coded dependencies. 
Regarding the taxonomy of the connections, i.e. the syntax of the returned connection object, I think it's the catalog connector that defines it, and the rest of components (modules, storage provider) should obey it. The connectors and the modules that are a part of "Fybrik open-source ecosystem" are all in sync. However, it requires the module developers to be aware of the connection taxonomy, while Fybrik does not provide/publish a default taxonomy object which may be a problem for the future module developers.

**Comment by roytman** [comment]
> Regarding the taxonomy of the connections, i.e. the syntax of the returned connection object, I think it's the catalog connector that defines it, and the rest of components (modules, storage provider) should obey it. 

Ideally, it is correct, but it is not always possible, so maybe some of the modules can support helpers that can convert one connection taxonomy to another.
